src/libraries/QueryResponse.sol

```diff
@@ -0,0 +1,590 @@
+  bytes response;
+}
+
+struct EthCallQueryResponse {
+  bytes requestBlockId;
+  uint64 blockNum;
+  uint64 blockTime;
+  bytes32 blockHash;
+  EthCallRecord[] results;
+}
+
+struct EthCallByTimestampQueryResponse {
+  bytes requestTargetBlockIdHint;
+  bytes requestFollowingBlockIdHint;
+  uint64 requestTargetTimestamp;
+  uint64 targetBlockNum;
+  uint64 targetBlockTime;
+  uint64 followingBlockNum;
+  bytes32 targetBlockHash;
+  bytes32 followingBlockHash;
+  uint64 followingBlockTime;
+  EthCallRecord[] results;
+}
+
+struct EthCallWithFinalityQueryResponse {
+  bytes requestBlockId;
+  bytes requestFinality;
+  uint64 blockNum;
+  uint64 blockTime;
+  bytes32 blockHash;
+  EthCallRecord[] results;
+}
+
+struct EthCallRecord {
+  address contractAddress;
+  bytes callData;
+  bytes result;
+}
+
+struct SolanaAccountQueryResponse {
+  bytes requestCommitment;
+  uint64 requestMinContextSlot;
+  uint64 requestDataSliceOffset;
+  uint64 requestDataSliceLength;
+  uint64 slotNumber;
+  uint64 blockTime;
+  bytes32 blockHash;
+  SolanaAccountResult[] results;
+}
+
+struct SolanaAccountResult {
+  bytes32 account;
+  uint64 lamports;
+  uint64 rentEpoch;
+  bool executable;
+  bytes32 owner;
+  bytes data;
+}
+
+struct SolanaPdaQueryResponse {
+  bytes requestCommitment;
+  uint64 requestMinContextSlot;
+  uint64 requestDataSliceOffset;
+  uint64 requestDataSliceLength;
+  uint64 slotNumber;
+  uint64 blockTime;
+  bytes32 blockHash;
+  SolanaPdaResult[] results;
+}
+
+struct SolanaPdaResult {
+  bytes32 programId;
+  bytes[] seeds;
+  bytes32 account;
+  uint64 lamports;
+  uint64 rentEpoch;
+  bool executable;
+  bytes32 owner;
+  bytes data;
+  uint8 bump;
+}
+
+error WrongQueryType(uint8 received, uint8 expected);
+error InvalidResponseVersion();
+error VersionMismatch();
+error ZeroQueries();
+error NumberOfResponsesMismatch();
+error ChainIdMismatch();
+error RequestTypeMismatch();
+error UnexpectedNumberOfResults();
+error InvalidPayloadLength(uint256 received, uint256 expected);
+error InvalidContractAddress();
+error InvalidFunctionSignature();
+error InvalidChainId();
+error StaleBlockNum();
+error StaleBlockTime();
+error VerificationFailed();
+
+//QueryResponse is a library that implements the parsing and verification of
+//  Cross Chain Query (CCQ) responses.
+//
+//For a detailed discussion of these query responses, please see the white paper:
+//  https://github.com/wormhole-foundation/wormhole/blob/main/whitepapers/0013_ccq.md
+library QueryResponseLib {
+  using BytesParsing for bytes;
+
+  bytes internal constant RESPONSE_PREFIX = bytes("query_response_0000000000000000000|");
+  uint8 internal constant VERSION = 1;
+  uint64 internal constant MICROSECONDS_PER_SECOND = 1_000_000;
+
+  function calcPrefixedResponseHash(bytes memory response) internal pure returns (bytes32) {
+    return keccak256(abi.encodePacked(RESPONSE_PREFIX, keccak256(response)));
+  }
+
+  //TODO add calldata impl? (duplicated code but better gas efficiency)
```

Comment:
Is that an open question for this PR?

Reply:
Sorta. I left it in there as a reminder and a comment, primarily.

The issue is that I don't really know how this is used. Introducing the `calldata` location essentially doubles the code size (because we definitely have to provide the `memory` version) of the library and it's not clear to me whether that's worth it.

It might encourage users to ask for the `calldata` version (or at least make them aware that maybe they'd want one).